### ISSUE TITLE
dlSym cannot locate method 'CreateExtension'

### ISSUE THREAD

**Issue by zhaohb** [triton-inference-server/server#3134]
**Triton Information**
What version of Triton are you using?  21.06

Are you using the Triton container or did you build it yourself?
build my self

I have now recompiled the openvino back using openvino 2021.2 and need to load a custom operator that has been compiled to so lib,  but got an error when loading so:

```shell
+---------+---------+------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Model   | Version | Status                                                                                                                                                           |
+---------+---------+------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| mylayer | 1       | UNAVAILABLE: Internal: ModelState::Create InferenceEngineException: dlSym cannot locate method 'CreateExtension': /data/custom_op_openvino/build/libt |
|         |         | emplate_extension.so: undefined symbol: CreateExtension                                                                                                          |
+---------+---------+------------------------------------------------------------------------------------------------------------------------------------------------------------------+

```

how to solve it?

**Comment by tanmayv25** [comment]
@zhaohb  Can you share the complete logs? Specifically, do you see This message being printed? 
https://github.com/triton-inference-server/openvino_backend/blob/main/src/openvino.cc#L288

If not, then can you try adding second "CPU" argument as described here: https://docs.openvinotoolkit.org/2021.2/openvino_docs_HOWTO_Custom_Layers_Guide.html#building_and_running_the_custom_extension
in the function call here: 
https://github.com/triton-inference-server/openvino_backend/blob/main/src/openvino.cc#L283
?

**Comment by zhaohb** [comment]
@tanmayv25 thank you very much for your help, now I have solved the problem about CreateExtension, because I need to add the InferenceEngine::CreateExtension interface to the custom operator, but there is an error when loading the model:
```shell
root@137f7746fc9d:/opt/tritonserver/backends/openvino# tritonserver --model-repository=repo --strict-model-config=false --log-verbose=0 --metrics-port=6000
Error: Failed to initialize NVML
W0721 12:30:48.640710 141 metrics.cc:214] DCGM unable to start: DCGM initialization error
I0721 12:30:49.050402 141 libtorch.cc:987] TRITONBACKEND_Initialize: pytorch
I0721 12:30:49.050440 141 libtorch.cc:997] Triton TRITONBACKEND API version: 1.4
I0721 12:30:49.050451 141 libtorch.cc:1003] 'pytorch' TRITONBACKEND API version: 1.4
2021-07-21 12:30:49.281147: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
I0721 12:30:49.347246 141 tensorflow.cc:2165] TRITONBACKEND_Initialize: tensorflow
I0721 12:30:49.347284 141 tensorflow.cc:2175] Triton TRITONBACKEND API version: 1.4
I0721 12:30:49.347295 141 tensorflow.cc:2181] 'tensorflow' TRITONBACKEND API version: 1.4
I0721 12:30:49.347305 141 tensorflow.cc:2205] backend configuration:
{}
I0721 12:30:49.348989 141 onnxruntime.cc:1969] TRITONBACKEND_Initialize: onnxruntime
I0721 12:30:49.349019 141 onnxruntime.cc:1979] Triton TRITONBACKEND API version: 1.4
I0721 12:30:49.349028 141 onnxruntime.cc:1985] 'onnxruntime' TRITONBACKEND API version: 1.4
I0721 12:30:49.397309 141 openvino.cc:1188] TRITONBACKEND_Initialize: openvino
I0721 12:30:49.397353 141 openvino.cc:1198] Triton TRITONBACKEND API version: 1.4
I0721 12:30:49.397369 141 openvino.cc:1204] 'openvino' TRITONBACKEND API version: 1.4
W0721 12:30:49.397531 141 pinned_memory_manager.cc:236] Unable to allocate pinned system memory, pinned memory pool will not be available: CUDA driver version is insufficient for CUDA runtime version
I0721 12:30:49.397586 141 cuda_memory_manager.cc:115] CUDA memory pool disabled
E0721 12:30:49.400053 141 logging.cc:43] CUDA initialization failure with error 35. Please check your CUDA installation:  http://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html
E0721 12:30:49.400122 141 logging.cc:43] CUDA initialization failure with error 35. Please check your CUDA installation:  http://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html
W0721 12:30:49.400142 141 autofill.cc:225] Autofiller failed to detect the platform for mylayer (verify contents of model directory or use --log-verbose=1 for more details)
W0721 12:30:49.400157 141 autofill.cc:243] Proceeding with simple config for now
I0721 12:30:49.400290 141 model_repository_manager.cc:1045] loading: mylayer:1
I0721 12:30:49.500777 141 openvino.cc:1243] TRITONBACKEND_ModelInitialize: mylayer (version 1)
W0721 12:30:49.504520 141 openvino.cc:632] skipping model configuration auto-complete for 'mylayer': not supported for openvino backend
I0721 12:30:49.505403 141 openvino.cc:188] model configuration:
{
    "name": "mylayer",
    "platform": "",
    "backend": "openvino",
    "version_policy": {
        "latest": {
            "num_versions": 1
        }
    },
    "max_batch_size": 0,
    "input": [
        {
            "name": "Parameter_0",
            "data_type": "TYPE_FP32",
            "format": "FORMAT_NONE",
            "dims": [
                1,
                3,
                10,
                11
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        }
    ],
    "output": [
        {
            "name": "Relu_3",
            "data_type": "TYPE_FP32",
            "dims": [
                1,
                3,
                10,
                11
            ],
            "label_filename": "",
            "is_shape_tensor": false
        }
    ],
    "batch_input": [],
    "batch_output": [],
    "optimization": {
        "priority": "PRIORITY_DEFAULT",
        "input_pinned_memory": {
            "enable": true
        },
        "output_pinned_memory": {
            "enable": true
        },
        "gather_kernel_buffer_threshold": 0,
        "eager_batching": false
    },
    "instance_group": [
        {
            "name": "mylayer_0",
            "kind": "KIND_CPU",
            "count": 1,
            "gpus": [],
            "secondary_devices": [],
            "profile": [],
            "passive": false,
            "host_policy": ""
        }
    ],
    "default_model_filename": "",
    "cc_model_filenames": {},
    "metric_tags": {},
    "parameters": {
        "CPU_EXTENSION_PATH": {
            "string_value": "/data/custom_ov_op/mylayer_lib/build/libmylayer_cpu_extension.so"
        }
    },
    "model_warmup": []
}
I0721 12:30:49.506001 141 openvino.cc:1288] TRITONBACKEND_ModelInstanceInitialize: mylayer_0 (CPU device 0)
I0721 12:30:49.517215 141 openvino.cc:286] CPU (MKLDNN) extensions is loaded/data/zhaohongbo/custom_ov_op/mylayer_lib/build/libmylayer_cpu_extension.so
I0721 12:30:49.545726 141 openvino.cc:1322] TRITONBACKEND_ModelInstanceFinalize: delete instance state
I0721 12:30:49.545760 141 openvino.cc:1268] TRITONBACKEND_ModelFinalize: delete model state
E0721 12:30:49.546962 141 model_repository_manager.cc:1215] failed to load 'mylayer' version 1: Internal: openvino error in loading network : Unsupported primitive of type: MyLayer name: MyLayer_5
I0721 12:30:49.547146 141 server.cc:504]
+------------------+------+
| Repository Agent | Path |
+------------------+------+
+------------------+------+

I0721 12:30:49.547399 141 server.cc:543]
+-------------+-----------------------------------------------------------------+--------+
| Backend     | Path                                                            | Config |
+-------------+-----------------------------------------------------------------+--------+
| tensorrt    | <built-in>                                                      | {}     |
| pytorch     | /opt/tritonserver/backends/pytorch/libtriton_pytorch.so         | {}     |
| tensorflow  | /opt/tritonserver/backends/tensorflow1/libtriton_tensorflow1.so | {}     |
| onnxruntime | /opt/tritonserver/backends/onnxruntime/libtriton_onnxruntime.so | {}     |
| openvino    | /opt/tritonserver/backends/openvino/libtriton_openvino.so       | {}     |
+-------------+-----------------------------------------------------------------+--------+

I0721 12:30:49.547558 141 server.cc:586]
+---------+---------+---------------------------------------------------------------------------------+
| Model   | Version | Status                                                                          |
+---------+---------+---------------------------------------------------------------------------------+
| mylayer | 1       | UNAVAILABLE: Internal: openvino error in loading network : Unsupported primitiv |
|         |         | e of type: MyLayer name: MyLayer_5                                              |
+---------+---------+---------------------------------------------------------------------------------+

I0721 12:30:49.547841 141 tritonserver.cc:1718]
+----------------------------------+------------------------------------------------------------------+
| Option                           | Value                                                            |
+----------------------------------+------------------------------------------------------------------+
| server_id                        | triton                                                           |
| server_version                   | 2.11.0                                                           |
| server_extensions                | classification sequence model_repository model_repository(unload |
|                                  | _dependents) schedule_policy model_configuration system_shared_m |
|                                  | emory cuda_shared_memory binary_tensor_data statistics           |
| model_repository_path[0]         | repo                               |
| model_control_mode               | MODE_NONE                                                        |
| strict_model_config              | 0                                                                |
| pinned_memory_pool_byte_size     | 268435456                                                        |
| min_supported_compute_capability | 6.0                                                              |
| strict_readiness                 | 1                                                                |
| exit_timeout                     | 30                                                               |
+----------------------------------+------------------------------------------------------------------+

I0721 12:30:49.547935 141 server.cc:234] Waiting for in-flight requests to complete.
I0721 12:30:49.547967 141 server.cc:249] Timeout 30: Found 0 live models and 0 in-flight non-inference requests
error: creating server: Internal - failed to load all models
```

model.xml with netron:
![image](https://user-images.githubusercontent.com/5791375/126489218-71fab229-bf17-44e7-8770-521f61547e17.png)

how to solve?

**Comment by zhaohb** [comment]
this is my mylayer.cpp:
```c++
#include <ngraph/ngraph.hpp>
#include <inference_engine.hpp>

// nGraph Op definition for custom layer
class MyLayerOp : public ngraph::op::Op {
public:
    static constexpr ngraph::NodeTypeInfo type_info{"MyLayer", 0};
    const ngraph::NodeTypeInfo& get_type_info() const override { return type_info;  }

    MyLayerOp() = default;

    MyLayerOp(const ngraph::Output<ngraph::Node>& input) : Op({input}) {
        constructor_validate_and_infer_types();
    }

    void validate_and_infer_types() override {
        auto inpShape = get_input_partial_shape(0);
        auto outShape = inpShape;
        set_output_type(0, get_input_element_type(0), outShape);
    }

    std::shared_ptr<ngraph::Node> clone_with_new_inputs(const ngraph::OutputVector& new_args) const override {
        return std::make_shared<MyLayerOp>(new_args.at(0));
    }

    bool visit_attributes(ngraph::AttributeVisitor& visitor) override {
        return true;
    }
};

constexpr ngraph::NodeTypeInfo MyLayerOp::type_info;

// Custom layer implementation
class MyLayerImpl : public InferenceEngine::ILayerExecImpl
{
public:
    explicit MyLayerImpl(const std::shared_ptr<ngraph::Node>& node)
    {
        inpShape = node->get_input_shape(0);
        outShape = node->get_output_shape(0);
    }

    ~MyLayerImpl()
    {
        // nothing
    }

    InferenceEngine::StatusCode init(InferenceEngine::LayerConfig& config,
                                     InferenceEngine::ResponseDesc *resp) noexcept
    {
        return InferenceEngine::StatusCode::OK;
    }

    virtual InferenceEngine::StatusCode
    getSupportedConfigurations(std::vector<InferenceEngine::LayerConfig>& conf,
                               InferenceEngine::ResponseDesc* resp) noexcept
    {
        std::vector<InferenceEngine::DataConfig> inDataConfig;
        std::vector<InferenceEngine::DataConfig> outDataConfig;
        InferenceEngine::SizeVector order(inpShape.size());
        std::iota(order.begin(), order.end(), 0);

        // Allow any offset before data
        size_t offset((std::numeric_limits<size_t>::max)());

        // Input shape
        InferenceEngine::DataConfig inpConf;
        inpConf.desc = InferenceEngine::TensorDesc(InferenceEngine::Precision::FP32, inpShape, {inpShape, order, offset});
        inDataConfig.push_back(inpConf);

        // Output shape
        InferenceEngine::DataConfig outConf;
        outConf.desc = InferenceEngine::TensorDesc(InferenceEngine::Precision::FP32, outShape, {outShape, order, offset});
        outDataConfig.push_back(outConf);

        InferenceEngine::LayerConfig layerConfig;
        layerConfig.inConfs = inDataConfig;
        layerConfig.outConfs = outDataConfig;

        conf.push_back(layerConfig);
        return InferenceEngine::StatusCode::OK;
    }

    virtual InferenceEngine::StatusCode execute(std::vector<InferenceEngine::Blob::Ptr>& inputs,
                                                std::vector<InferenceEngine::Blob::Ptr>& outputs,
                                                InferenceEngine::ResponseDesc *resp) noexcept
    {
        const float* inp = inputs[0]->cbuffer().as<float*>();
        float* out = outputs[0]->buffer().as<float*>();
        for (size_t i = 0; i < inputs[0]->size(); ++i) {
            out[i] = inp[i] + 1.0f;
        }
        return InferenceEngine::OK;
    }

private:
    ngraph::Shape inpShape;
    ngraph::Shape outShape;
};


class InfEngineNgraphExtension : public InferenceEngine::IExtension
{
public:
    void Unload() noexcept override {}
    void Release() noexcept override { delete this; }
    void GetVersion(const InferenceEngine::Version*& versionInfo) const noexcept override {
	        static InferenceEngine::Version ExtensionDescription = {
        {1, 0},           // extension API version
        "1.0",
        "template_ext"    // extension description message
    };

    versionInfo = &ExtensionDescription;
    }
    std::vector<std::string> getImplTypes(const std::shared_ptr<ngraph::Node>& node) override {
        return {"CPU"};
    }

    InferenceEngine::ILayerImpl::Ptr getImplementation(const std::shared_ptr<ngraph::Node>& node,
                                                       const std::string& implType) override {
        if (std::dynamic_pointer_cast<MyLayerOp>(node) && implType == "CPU") {
            return std::make_shared<MyLayerImpl>(node);
        }
        return nullptr;
    }
};

#if 1
INFERENCE_EXTENSION_API(InferenceEngine::StatusCode) InferenceEngine::CreateExtension(InferenceEngine::IExtension *&ext,
                                                                                      InferenceEngine::ResponseDesc *resp) noexcept {
    try {
        ext = new InfEngineNgraphExtension();
        return OK;
    } catch (std::exception &ex) {
        if (resp) {
            std::string err = ((std::string) "Couldn't create extension: ") + ex.what();
            err.copy(resp->msg, 255);
        }
        return InferenceEngine::GENERAL_ERROR;
    }
}
#endif
```

**Comment by tanmayv25** [comment]
> I have solved the problem about CreateExtension, because I need to add the InferenceEngine::CreateExtension interface to the custom operator,

Great you found this issue!

> Internal: openvino error in loading network : Unsupported primitive of type: MyLayer name: MyLayer_5

This issue might be better answered by  Intel OpenVINO (https://github.com/openvinotoolkit/openvino). 
I wonder why MyLayer is still being searched as a primitive type. 

Unfortunately, we don't have much testing of these features on the backend yet.

**Comment by zhaohb** [comment]
I think we should implement a mylayer.py in openvino/deployment_tools/model_optimizer/extensions/ops , like this:
```python
class MyLayer(Op):
    op = 'MyLayer'
    enabled = True

    def __init__(self, graph: Graph, attrs: dict):
        super().__init__(graph, {
            'type': self.op,
            'op': self.op,
            'version': 'extension',
            'inverse': None,
            'in_ports_count': 1,
            'out_ports_count': 1,
            'infer': copy_shape_infer
        }, attrs)

    def backend_attrs(self):
        return []
```

but in openvino_backend I do not know where is be.

**Comment by zhaohb** [comment]
the next version of openvino backend will support custom op？I need openvino to support the function of custom op.

**Comment by zhaohb** [comment]
@tanmayv25 I implement getOpSets in InfEngineNgraphExtension class：
```c++
    std::map<std::string, ngraph::OpSet> getOpSets() {
        std::map<std::string, ngraph::OpSet> opsets;
        ngraph::OpSet opset;
        opset.insert<MyLayerOp>();
        opsets["extension"] = opset;
        return opsets;
    }
```

but I get the error:
```shell
Error: Failed to initialize NVML
W0722 09:30:04.947087 754 metrics.cc:214] DCGM unable to start: DCGM initialization error
I0722 09:30:05.370086 754 libtorch.cc:987] TRITONBACKEND_Initialize: pytorch
I0722 09:30:05.370120 754 libtorch.cc:997] Triton TRITONBACKEND API version: 1.4
I0722 09:30:05.370130 754 libtorch.cc:1003] 'pytorch' TRITONBACKEND API version: 1.4
2021-07-22 09:30:05.614239: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
I0722 09:30:05.666457 754 tensorflow.cc:2165] TRITONBACKEND_Initialize: tensorflow
I0722 09:30:05.666491 754 tensorflow.cc:2175] Triton TRITONBACKEND API version: 1.4
I0722 09:30:05.666504 754 tensorflow.cc:2181] 'tensorflow' TRITONBACKEND API version: 1.4
I0722 09:30:05.666514 754 tensorflow.cc:2205] backend configuration:
{}
I0722 09:30:05.668244 754 onnxruntime.cc:1969] TRITONBACKEND_Initialize: onnxruntime
I0722 09:30:05.668274 754 onnxruntime.cc:1979] Triton TRITONBACKEND API version: 1.4
I0722 09:30:05.668284 754 onnxruntime.cc:1985] 'onnxruntime' TRITONBACKEND API version: 1.4
I0722 09:30:05.709673 754 openvino.cc:1188] TRITONBACKEND_Initialize: openvino
I0722 09:30:05.709706 754 openvino.cc:1198] Triton TRITONBACKEND API version: 1.4
I0722 09:30:05.709722 754 openvino.cc:1204] 'openvino' TRITONBACKEND API version: 1.4
W0722 09:30:05.709883 754 pinned_memory_manager.cc:236] Unable to allocate pinned system memory, pinned memory pool will not be available: CUDA driver version is insufficient for CUDA runtime version
I0722 09:30:05.709933 754 cuda_memory_manager.cc:115] CUDA memory pool disabled
E0722 09:30:05.712487 754 logging.cc:43] CUDA initialization failure with error 35. Please check your CUDA installation:  http://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html
E0722 09:30:05.712560 754 logging.cc:43] CUDA initialization failure with error 35. Please check your CUDA installation:  http://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html
W0722 09:30:05.712580 754 autofill.cc:225] Autofiller failed to detect the platform for mylayer (verify contents of model directory or use --log-verbose=1 for more details)
W0722 09:30:05.712595 754 autofill.cc:243] Proceeding with simple config for now
I0722 09:30:05.712701 754 model_repository_manager.cc:1045] loading: mylayer:1
I0722 09:30:05.813198 754 openvino.cc:1243] TRITONBACKEND_ModelInitialize: mylayer (version 1)
W0722 09:30:05.816965 754 openvino.cc:632] skipping model configuration auto-complete for 'mylayer': not supported for openvino backend
I0722 09:30:05.817841 754 openvino.cc:188] model configuration:
{
    "name": "mylayer",
    "platform": "",
    "backend": "openvino",
    "version_policy": {
        "latest": {
            "num_versions": 1
        }
    },
    "max_batch_size": 0,
    "input": [
        {
            "name": "Parameter_0",
            "data_type": "TYPE_FP32",
            "format": "FORMAT_NONE",
            "dims": [
                1,
                3,
                10,
                11
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        }
    ],
    "output": [
        {
            "name": "Relu_3",
            "data_type": "TYPE_FP32",
            "dims": [
                1,
                3,
                10,
                11
            ],
            "label_filename": "",
            "is_shape_tensor": false
        }
    ],
    "batch_input": [],
    "batch_output": [],
    "optimization": {
        "priority": "PRIORITY_DEFAULT",
        "input_pinned_memory": {
            "enable": true
        },
        "output_pinned_memory": {
            "enable": true
        },
        "gather_kernel_buffer_threshold": 0,
        "eager_batching": false
    },
    "instance_group": [
        {
            "name": "mylayer_0",
            "kind": "KIND_CPU",
            "count": 1,
            "gpus": [],
            "secondary_devices": [],
            "profile": [],
            "passive": false,
            "host_policy": ""
        }
    ],
    "default_model_filename": "",
    "cc_model_filenames": {},
    "metric_tags": {},
    "parameters": {
        "CPU_EXTENSION_PATH": {
            "string_value": "/data/zhaohongbo/custom_ov_op/mylayer_lib/build/libmylayer_cpu_extension.so"
        }
    },
    "model_warmup": []
}
I0722 09:30:05.818429 754 openvino.cc:1288] TRITONBACKEND_ModelInstanceInitialize: mylayer_0 (CPU device 0)
I0722 09:30:05.826187 754 openvino.cc:1322] TRITONBACKEND_ModelInstanceFinalize: delete instance state
I0722 09:30:05.826258 754 openvino.cc:1268] TRITONBACKEND_ModelFinalize: delete model state
E0722 09:30:05.826470 754 model_repository_manager.cc:1215] failed to load 'mylayer' version 1: Internal: openvino error in reading network : Cannot create MyLayer layer MyLayer_5 id:1 from unsupported opset: extension
I0722 09:30:05.826637 754 server.cc:504]
+------------------+------+
| Repository Agent | Path |
+------------------+------+
+------------------+------+

I0722 09:30:05.826909 754 server.cc:543]
+-------------+-----------------------------------------------------------------+--------+
| Backend     | Path                                                            | Config |
+-------------+-----------------------------------------------------------------+--------+
| tensorrt    | <built-in>                                                      | {}     |
| pytorch     | /opt/tritonserver/backends/pytorch/libtriton_pytorch.so         | {}     |
| tensorflow  | /opt/tritonserver/backends/tensorflow1/libtriton_tensorflow1.so | {}     |
| onnxruntime | /opt/tritonserver/backends/onnxruntime/libtriton_onnxruntime.so | {}     |
| openvino    | /opt/tritonserver/backends/openvino/libtriton_openvino.so       | {}     |
+-------------+-----------------------------------------------------------------+--------+

I0722 09:30:05.827122 754 server.cc:586]
+---------+---------+---------------------------------------------------------------------------------+
| Model   | Version | Status                                                                          |
+---------+---------+---------------------------------------------------------------------------------+
| mylayer | 1       | UNAVAILABLE: Internal: openvino error in reading network : Cannot create MyLaye |
|         |         | r layer MyLayer_5 id:1 from unsupported opset: extension                        |
+---------+---------+---------------------------------------------------------------------------------+

I0722 09:30:05.827510 754 tritonserver.cc:1718]
+----------------------------------+------------------------------------------------------------------+
| Option                           | Value                                                            |
+----------------------------------+------------------------------------------------------------------+
| server_id                        | triton                                                           |
| server_version                   | 2.11.0                                                           |
| server_extensions                | classification sequence model_repository model_repository(unload |
|                                  | _dependents) schedule_policy model_configuration system_shared_m |
|                                  | emory cuda_shared_memory binary_tensor_data statistics           |
| model_repository_path[0]         | /data/zhaohongbo/custom_ov_op/repo/                              |
| model_control_mode               | MODE_NONE                                                        |
| strict_model_config              | 0                                                                |
| pinned_memory_pool_byte_size     | 268435456                                                        |
| min_supported_compute_capability | 6.0                                                              |
| strict_readiness                 | 1                                                                |
| exit_timeout                     | 30                                                               |
+----------------------------------+------------------------------------------------------------------+

I0722 09:30:05.827609 754 server.cc:234] Waiting for in-flight requests to complete.
I0722 09:30:05.827647 754 server.cc:249] Timeout 30: Found 0 live models and 0 in-flight non-inference requests
error: creating server: Internal - failed to load all models
```
model.xml like this:
```xml
<?xml version="1.0"?>
<net name="Function_0" version="10">
	<layers>
		<layer id="0" name="Parameter_0" type="Parameter" version="opset1">
			<data shape="1, 3, 10, 11" element_type="f32" />
			<output>
				<port id="0" precision="FP32">
					<dim>1</dim>
					<dim>3</dim>
					<dim>10</dim>
					<dim>11</dim>
				</port>
			</output>
		</layer>
		<layer id="1" name="MyLayer_5" type="MyLayer" version="extension">
			<input>
				<port id="0">
					<dim>1</dim>
					<dim>3</dim>
					<dim>10</dim>
					<dim>11</dim>
				</port>
			</input>
			<output>
				<port id="1" precision="FP32">
					<dim>1</dim>
					<dim>3</dim>
					<dim>10</dim>
					<dim>11</dim>
				</port>
			</output>
		</layer>
		<layer id="2" name="Relu_3" type="ReLU" version="opset1">
			<input>
				<port id="0">
					<dim>1</dim>
					<dim>3</dim>
					<dim>10</dim>
					<dim>11</dim>
				</port>
			</input>
			<output>
				<port id="1" precision="FP32">
					<dim>1</dim>
					<dim>3</dim>
					<dim>10</dim>
					<dim>11</dim>
				</port>
			</output>
		</layer>
		<layer id="3" name="Result_4" type="Result" version="opset1">
			<input>
				<port id="0">
					<dim>1</dim>
					<dim>3</dim>
					<dim>10</dim>
					<dim>11</dim>
				</port>
			</input>
		</layer>
	</layers>
	<edges>
		<edge from-layer="0" from-port="0" to-layer="1" to-port="0" />
		<edge from-layer="1" from-port="1" to-layer="2" to-port="0" />
		<edge from-layer="2" from-port="1" to-layer="3" to-port="0" />
	</edges>
</net>
```

It gives me the feeling that the custom op is not loaded.

**Comment by tanmayv25** [comment]
> I0721 12:30:49.517215 141 openvino.cc:286] CPU (MKLDNN) extensions is loaded/data/zhaohongbo/custom_ov_op/mylayer_lib/build/libmylayer_cpu_extension.so

The above message doesn't seems to be getting emitted with your `getOpSets` implementation. In fact, the error seems to be coming during reading the IR itself. If you are able to use your extension outside Triton, in OV environment, then it is definitely a bug with our backend implementation. Can you share a minimal repro with your *final* extension that works within OV but fails with Triton? 
We will try to fix it when we get to it. Otherwise, if you are able to find a bug with Triton OV backend implementation then feel free to raise a PR.

**Comment by zhaohb** [comment]
@tanmayv25 [https://drive.google.com/drive/folders/1RPM2PjFMJaSdwDX-cpQ3gsqpTQ8B0fZf?usp=sharing](https://drive.google.com/drive/folders/1RPM2PjFMJaSdwDX-cpQ3gsqpTQ8B0fZf?usp=sharing) 
you can get code from there.

1. mylayer.tar.gz is the code that can be executed correctly in the ov environment, it will generate the ngraph_custom_layer executable program, and he will generate the ov model model xml; the steps of compilation : 1).cd build  2).cmake ..  3). make

2. mylayer_lib.tar.gz is the code for compiling the so library on the ov environment, it will generate libmylayer_cpu_extension.so; the steps of compilation : 1).cd build  2).cmake ..  3). make

3. repo.tar.gz is the model repo of triton. had add libmylayer_cpu_extension.so to config.pbtxt, modify the path of libmylayer_cpu_extension.so according to the actual situation, and use triton to load the model will reproduce the problem


I hope it can be resolved as soon as possible, and I will continue to follow up, thanks very much.

**Comment by zhaohb** [comment]
@tanmayv25 did you repeat the problem?

**Comment by tanmayv25** [comment]
Not yet. Have been occupied with some other higher priority tasks. I will try to get on this next week.

**Comment by zhaohb** [comment]
@tanmayv25 I have solved the problem. and will submit a patch.

**Comment by tanmayv25** [comment]
A patch will be greatly appreciated! Please send a CLA agreement as specified [here](https://github.com/triton-inference-server/server/blob/main/CONTRIBUTING.md#contributor-license-agreement-cla).

**Comment by zhaohb** [comment]
@tanmayv25  I think this bug can be fixed in 2021.4, then I don't have to compile it myself. is it feasible？

**Comment by zhaohb** [comment]
I had push my change   [https://github.com/triton-inference-server/openvino_backend/pull/27](https://github.com/triton-inference-server/openvino_backend/pull/27)

**Comment by zhaohb** [comment]
@tanmayv25 hello, now  I had recompiled openvino_backend, and inference my model, but I found that openvino's throughput is much less and latency is much higher than TensorFlow. the step compile:

- $ mkdir build
- $ cd build
- $cmake - DCMAKE_INSTALL_PREFIX: PATH = ` PWD ` / install - DTRITON_BUILD_OPENVINO_VERSION = 2021.3 - DTRITON_BUILD_CONTAINER_VERSION = 20.16 ..
- make 
- make install

Do I need to add some special options to use openvino better? do you have any suggest?

**Comment by zhaohb** [comment]
I had compiled TensorFlow backend before, and without changing the code,, but the performance was also significantly reduced, I do not know why.

**Comment by tanmayv25** [comment]
> Do I need to add some special options to use openvino better? do you have any suggest?

You can experiment with different `CPU_THREADS_NUM`, `CPU_BIND_THREAD` and `CPU_THROUGHPUT_STREAMS` settings. More details on these parameters [here](https://github.com/triton-inference-server/openvino_backend#parameters) and [here](https://docs.openvinotoolkit.org/latest/openvino_docs_IE_DG_supported_plugins_CPU.html#supported_configuration_parameters). In our experiments we found the performance to be sensitive for these parameters.

> get libMKLDNNPlugin.so from deb.

Why? The backend build script should build the plugin for you. The r21.06 should build openVINO with OMP support whereas release debs use intel TBB.

**Comment by tanmayv25** [comment]
> I had compiled TensorFlow backend before, and without changing the code,, but the performance was also significantly reduced, I do not know why.

May be something changed in the system?

**Comment by zhaohb** [comment]
> Do I need to add some special options to use openvino better? do you have any suggest?

> > You can experiment with different CPU_THREADS_NUM, CPU_BIND_THREAD and CPU_THROUGHPUT_STREAMS settings. More details on these parameters here and here. In our experiments we found the performance to be sensitive for these parameters.

> get libMKLDNNPlugin.so from deb.

> > Why? The backend build script should build the plugin for you. The r21.06 should build openVINO with OMP support whereas release debs use intel TBB.

I see tools/gen_openvino_dockerfile.py, 21.06 is not build with source, 21.07 will build with source.

**Comment by tanmayv25** [comment]
> could you help me debug the model? I can give you a repo including TF model and OV model. It's a recommendation model

Great! Can you open a new issue for perf tuning as your PR fixes the original issue?

 > I see tools/gen_openvino_dockerfile.py, 21.06 is not build with source, 21.07 will build with source.
 
 My bad. Thanks for correction.

**Comment by zhaohb** [comment]
> Great! Can you open a new issue for perf tuning as your PR fixes the original issue?

ok, I will open a new issue. and I will resubmit the PR as soon as possible

> My bad. Thanks for correction.
so, 21.06 use TBB？if I want get better perf, which I should use?openmp or tbb?

**Comment by tanmayv25** [comment]
[This ](https://docs.openvinotoolkit.org/latest/openvino_docs_optimization_guide_dldt_optimization_guide.html#cpu-checklist) is general guidance from openvino team. 

We moved from TBB to openMP because we found openMP version to give better perf with Triton. But our experiments have not been very exhaustive.

**Comment by zhaohb** [comment]
ok, if I want to use openmp, I should compiler OV with openvino_backend 21.07? but 21.07 has host policy, I do not why add it , We can also set it outside triton, for example in K8s.

**Comment by zhaohb** [comment]
I had open a new issue for perf tuning [https://github.com/triton-inference-server/server/issues/3194](https://github.com/triton-inference-server/server/issues/3194)

**Comment by tanmayv25** [comment]
Fixed by [#27](https://github.com/triton-inference-server/openvino_backend/pull/27)
